docs/adr/0007-seperate-binding-generators.md

```diff
@@ -0,0 +1,142 @@
+# Separate the binding generators into their own crates per target
+
+* Status: proposed
+* Deciders: TBD
+* Date: 2022-01-16
+
+Technical Story: [Issue 299](https://github.com/mozilla/uniffi-rs/issues/299)
+
+outdated early prototype: [PR 997](https://github.com/mozilla/uniffi-rs/pull/997)
+
+## Context and Problem Statement
+All the binding generators currently live in the [`uniffi_bindgen`](../../uniffi_bindgen/src/bindings) crate. This creates the following difficulties:
+- All the bindings live in the `uniffi` repository, so the `uniffi` team has to maintain them (or at the very least review changes to them).
+- Any changes to a specific binding generator require a new `uniffi_bindgen` release for it to be released. Even if it doesn't impact any of the other bindings.
+- Some bindings require complex build systems to test, and including those build systems in `uniffi` adds its complexity. For example, any type of `gecko-js` bindings would require the mozilla-central build system to build and test.
+- We currently run all the tests for the bindings in our CI and through `cargo test`. This means that if one binding target gets outdated and fails, or if a developer doesn't have the needed libraries installed for one of the targets, the tests would fail.
+
+This ADR re-evaluates the architecture of how we consume the bindings in hope of improving the developer experience of adding new bindings, testing new and the current bindings and simplifying the ownership of the bindings.
+## Decision Drivers
+
+* Maintainability, we would like to improve the maintainability of our binding generators - this means clear ownership of the bindings.
+* Testability, it should be easy for developers to test the bindings they care about. Without having to navigate and install unfamiliar libraries and build systems.
+* Version compatibility, we need to guarantee that the same version of `uniffi` is used to generate both the scaffolding and the bindings. This is easy to ensure when all the bindings (and the scaffolding) live under the same crate but becomes a concern if they are separated.
```

Comment:
We may be able to re-frame this constraint a little. IIRC, we don't *guarantee* this even now with everything in the same crate, because you have to invoke `uniffi-bindgen` once for the scaffolding and again for the bindings, and if you really try we probably can't stop you from using a different version of `uniffi-bindgen` for those two calls. I think the questions here are:

* How easy is it to accidentally generate scaffolding and bindings with different versions of UniFFI? (Probably this proposal will make it easier to do that accidentally).
* How much damage can mismatched UniFFI versions cause? (In theory, I think the checksum still prevents memory unsafety here, the failure mode is a hard build- or link-time error and this proposal wouldn't change that).
* How hard is it to debug what's gone wrong if this does happen? (Maybe this proposal is a chance to make it *easier* to debug?)

Reply:
We also have the option of relaxing the "must use the exact some version of UniFFI requirement" and replacing it with something finer-grained, such as "must use the same version of the UniFFI FFI Contract Specification (tm)".

The swift bindings [do a basic version of this](https://github.com/mozilla/uniffi-rs/blob/7dcbe1219eed36545effb29d8f30ca3425154212/uniffi_bindgen/src/bindings/swift/templates/BridgingHeaderTemplate.h#L15) but as you'll see in that file, it requires the presence of a comment block with ⚠️ signs in it warning folks that they must increment the version number when they make a breaking change.

Reply:
> We also have the option of relaxing the "must use the exact some version of UniFFI requirement" and replacing it with something finer-grained, such as "must use the same version of the UniFFI FFI Contract Specification (tm)".

This could be a really good follow-up to this work!! I also really like the three questions you framed so I'll steal those and rephrase them directly into the doc 😝